assets/src/models/createDetourMachine.ts

```diff
@@ -1,17 +1,139 @@
-import { setup } from "xstate"
+import { setup, assign, fromPromise } from "xstate"
+import { RouteId, RoutePattern, RoutePatternId } from "../schedule"
+import { fetchRoutePatterns } from "../api"
 
 export const createDetourMachine = setup({
   types: {} as {
-    events: { type: "detour.edit.done" } | { type: "detour.edit.resume" }
+    context: {
+      routeId?: RouteId
+      routePatternId?: RoutePatternId
+
+      routePatterns?: RoutePattern[]
+    }
+
+    input:
+      | {
+          // Caller has target route pattern
+          routeId: RouteId
+          routePatternId: RoutePatternId
+        }
+      | {
+          // Caller has target route
+          routeId: RouteId
+          routePatternId: undefined
+        }
+      | {
+          // Caller has no prior selection
+          routeId: undefined
+          routePatternId: undefined
+        }
+
+    events:
+      | { type: "detour.edit.done" }
+      | { type: "detour.edit.resume" }
+      | { type: "detour.route-pattern.open" }
+      | { type: "detour.route-pattern.done" }
+      | { type: "detour.route-pattern.select-route"; routeId: RouteId }
+      | {
+          type: "detour.route-pattern.select-pattern"
+          routePatternId: RoutePatternId
+        }
+  },
+  actors: {
+    "fetch-route-patterns": fromPromise<
+      Awaited<ReturnType<typeof fetchRoutePatterns>>,
+      { routeId?: RouteId }
+    >(async ({ input: { routeId } }) => {
+      if (routeId) {
+        return fetchRoutePatterns(routeId)
+      } else {
+        // Hmm, this could leave us in the lurch if we don't handle it,
+        // I'd like to make this impossible but I can't get typescript to infer
+        // the context types but it's not letting me
+        throw "No Route ID"
+      }
+    }),
+  },
+  actions: {
+    "set.route-pattern": assign({
+      routePatternId: (_, params: { routePatternId: RoutePatternId }) =>
+        params.routePatternId,
+    }),
+    "set.route-id": assign({
+      routeId: (_, params: { routeId: RouteId }) => params.routeId,
+    }),
   },
 }).createMachine({
   id: "Detours Machine",
+  context: ({ input }) => ({
+    ...input,
+  }),
 
   initial: "Detour Drawing",
   states: {
     "Detour Drawing": {
       initial: "Editing",
       states: {
+        "Pick Route Pattern": {
+          initial: "Pick Route ID",
+          on: {
+            "detour.route-pattern.select-route": {
+              target: ".Load Route Patterns",
```

Comment:
Is this a typo? (or a syntax I don't know yet?)

Reply:
Good question! This is valid syntax and not a typo.

See: 
> Targets the descendant `closed` state node
> -- https://stately.ai/docs/finite-states#targets